include/graph.h

```diff
@@ -1,138 +1,174 @@
-   *                   calling thread.
-   * @returns nothing (supernode delta is in delta_loc).
-   */
-  static void generate_delta_node(node_id_t node_n, long node_seed, node_id_t src,
-                                  const std::vector<size_t> &edges, Supernode *delta_loc);
-
-  /**
-   * Serialize the graph data to a binary file.
-   * @param filename the name of the file to (over)write data to.
-   */
-  void write_binary(const std::string &filename);
-
-  std::chrono::steady_clock::time_point flush_call;
-  std::chrono::steady_clock::time_point flush_return;
-  std::chrono::steady_clock::time_point cc_alg_start;
-  std::chrono::steady_clock::time_point cc_alg_end;
-  std::chrono::steady_clock::time_point create_backup_start;
-  std::chrono::steady_clock::time_point create_backup_end;
-  std::chrono::steady_clock::time_point restore_backup_start;
-  std::chrono::steady_clock::time_point restore_backup_end;
-};
-
+#pragma once
+#include <cstdlib>
+#include <exception>
+#include <set>
+#include <fstream>
+#include <atomic>  // REMOVE LATER
+
+#include <buffering_system.h>
+#include "supernode.h"
+
+#ifdef VERIFY_SAMPLES_F
+#include "test/graph_verifier.h"
+#endif
+
+// forward declarations
+class GraphWorker;
+
+typedef std::pair<Edge, UpdateType> GraphUpdate;
+
+// Exceptions the Graph class may throw
+class UpdateLockedException : public std::exception {
+  virtual const char* what() const throw() {
+    return "The graph cannot be updated: Connected components algorithm has "
+           "already started";
+  }
+};
+
+class MultipleGraphsException : public std::exception {
+  virtual const char * what() const throw() {
+    return "Only one Graph may be open at one time. The other Graph must be deleted.";
+  }
+};
+
+/**
+ * Undirected graph object with n nodes labelled 0 to n-1, no self-edges,
+ * multiple edges, or weights.
+ */
+class Graph {
+  node_id_t num_nodes;
+  long seed;
+  bool update_locked = false;
+  bool modified = false;
+  // a set containing one "representative" from each supernode
+  std::set<node_id_t>* representatives;
+  Supernode** supernodes;
+  // DSU representation of supernode relationship
+  node_id_t * parent;
+  node_id_t get_parent(node_id_t node);
+
+  // Buffering system for batching updates
+  BufferingSystem *bf;
+
+  void backup_to_disk(std::vector<node_id_t> ids_to_backup);
+  void restore_from_disk(std::vector<node_id_t> ids_to_restore);
+
+  /**
+   * Update the query array with new samples
+   * @param query  an array of supernode query results
+   * @param reps   an array containing node indices for the representative of each supernode
+   */
+  void sample_supernodes(
+    std::pair<Edge, SampleSketchRet> *query,
+    std::vector<node_id_t> &reps
+  );
+
+  /**
+   * @param copy_supernodes  an array to be filled with supernodes
+   * @param to_merge         an list of lists of supernodes to be merged
+   *
+   */
+  void merge_supernodes(
```

Comment:
Style nit: To make your function declarations consistent with the rest of the code, please have arguments on one line and make a newline if necessary for length. See generate_delta_node for an example of how we do this.